python/ray/tune/tests/test_ray_trial_executor.py

```diff
@@ -280,6 +282,59 @@ def create_trial(cpu, gpu=0):
             self.trial_executor.has_resources_for_trial(cpu_only_trial3))
 
 
+class TrialExecutorInsufficientResourceTest(unittest.TestCase):
+    def setUp(self):
+        os.environ["TUNE_INSUFFICENT_RESOURCE_WARN_THRESHOLD_S"] = "1"
+        self.cluster = Cluster(
+            initialize_head=True,
+            connect=True,
+            head_node_args={
+                "num_cpus": 4,
+                "num_gpus": 2,
+                "_system_config": {
+                    "num_heartbeats_timeout": 10
+                }
+            })
+        # Pytest doesn't play nicely with imports
+        _register_all()
+
+    def tearDown(self):
+        ray.shutdown()
+        self.cluster.shutdown()
+        _register_all()  # re-register the evicted objects
+
+    @freeze_time("2021-08-03", auto_tick_seconds=15)
+    def testOutputWarningMessage(self):
+        def train(config):
+            pass
+
+        with self.assertLogs(logger="ray.tune.trial_executor") as ctx:
+            out = tune.run(
```

Comment:
The set up in this file either drives the tests from `tune.run()` or on a granular level by interacting(start/stop/pause/resume) with individual trials. 

For me, the added logic is more or less hooked up with step(), thus driven by tune.run(). Is there a good way to drive heterogeneous trials? So that we can exercise the logic of:
- one trial runs and terminates successfully
- another trial requires more  resource than the 1st trial and thus we print the warning msg.

I am open to ideas to add more test coverage.

Reply:
@krfricke to see if any input for unit tests.